### ISSUE TITLE
Speedup for gauss functions

### ISSUE THREAD

**Issue by heckflosse** [Beep6581/RawTherapee#3014]
I'll post a PR with a speedup for gauss functions tomorrow.

Ingo

**Comment by heckflosse** [comment]
Still working an speedup and cleanup...

**Comment by heckflosse** [comment]
It turned out that I'll need a bit more time for this speedup but it will be also a larger speedup than expected (At current state of my changes RL sharpening a D800 file using RL 0.55 / 100 / 0 / 100 went down from 16318 to 9152 ms (and there is some room for additional speedup)

Ingo

**Comment by heckflosse** [comment]
Ok, here's the first branch https://github.com/Beep6581/RawTherapee/commits/rldeconv_speedup with my changes to gauss.h and some code cleanup for ipsharpen.cc
RL Sharpening a full D800 file using 0.55 /100/0/100 is reduced from 16.3 to 7.2 seconds at my system.
For RL radius >= 0.6 the speedup is much lower (in the range of about 20%)
When RL dampening is enabled (>0) the speedup is lower than without dampening but still serious.
The current branch is not complete (I've to code the non SSE2 version and clean the code a bit) but it should be good to allow tests on cpus which support SSE2.

Ingo

**Comment by Beep6581** [comment]
Compilation fails with:
```
[ 10%] Building CXX object rtengine/CMakeFiles/rtengine.dir/dcp.cc.o
/home/drslony/repo-rt/rtengine/ipsharpen.cc:97:6: error: prototype for ‘void rtengine::ImProcFunctions::deconvsharpening(float**, float**, int, int, const rtengine::procparams::SharpeningParams&)’ does not match any in class ‘rtengine::ImProcFunctions’
 void ImProcFunctions::deconvsharpening (float** luminance, float** tmp, int W, int H, const SharpeningParams &sharpenParam)
      ^
In file included from /home/drslony/repo-rt/rtengine/ipsharpen.cc:20:0:
/home/drslony/repo-rt/rtengine/improcfun.h:273:10: error: candidate is: void rtengine::ImProcFunctions::deconvsharpening(rtengine::LabImage*, float**, rtengine::procparams::SharpeningParams&)
     void deconvsharpening (LabImage* lab, float** buffer, SharpeningParams &sharpenParam);
          ^
/home/drslony/repo-rt/rtengine/ipsharpen.cc:121:25: error: expected ‘}’ before end of line
     #pragma omp parallel
                         ^
/home/drslony/repo-rt/rtengine/ipsharpen.cc:121:25: error: expected declaration before end of line
rtengine/CMakeFiles/rtengine.dir/build.make:1190: recipe for target 'rtengine/CMakeFiles/rtengine.dir/ipsharpen.cc.o' failed
```

**Comment by heckflosse** [comment]
Oups, forgot one file. Fixed.

**Comment by Beep6581** [comment]
I tested several files patched vs unpatched,
```
Method=rld
DeconvRadius=0.75
DeconvAmount=100
DeconvDamping=0
DeconvIterations=100
```

No differences. 25% speedup (down from 12s to 9s).

**Comment by adamreichold** [comment]
@heckflosse An impressive improvement!

From just reading through the code, I have two questions resp. suggestions:
* Is it strictly necessary to expose the implementation in the header? From similar situations with templated implementations, I remember the pattern of using
````
void doSomething(float x);
void doSomething(double x);
````
in the header and then something like
````
namespace
{
template<typename T>
void doSomethingImpl(T x)
{
    /* ... */
}
}
void doSomething(float x) { doSomethingImpl(x); }
void doSomething(double x) { doSomethingImpl(x); }
````
in a separated compiled source. 
* DId you try using OpenMP 4.0 [SIMD pragmas](https://software.intel.com/en-us/articles/enabling-simd-in-program-using-openmp40) instead of manually unrolling those loops? It could be that it is not as fast as a hand-coded implementation using intrinsics, but maybe it is easier to experiment with different amounts of unrolling and/or threading versus vectorization.

**Comment by heckflosse** [comment]
@adamreichold : gcc less than version 4.9 does not support OpenMP 4.0. As long as we use gcc < 4.9 to build Win32 RT, using OpenMP 4.0 features is not an option.

It's not necessary to expose the template implementation in the header. It's just that I didn't care about it. But changing it is a good idea and I'll do that with the next commit.

Ingo

**Comment by adamreichold** [comment]
@heckflosse Too bad about OpenMP 4.0, so intrinsics it is. Moving the implementation into a separate source file should be useful, since it will decrease build times and it can also lead to more a compact executable since the implementation is not duplicated (and most linkers will not remove the duplicate implementations, especially without LTO), which can for example yield better cache utilization.

**Comment by heckflosse** [comment]
I pushed some changes to https://github.com/Beep6581/RawTherapee/tree/rldeconv_speedup.
@adamreichold As suggested by you I split gauss.h into gauss.cc and gauss.h (very good idea)
Because of that split I had to touch some files. During that work I could reduce processing time of Lab Impulse Noise reduction by 25%. I also included some code cleanups.
I tested the gauss stuff using SSE and non-SSE build. Seems to work fine here.

Ingo
